features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineDnsCache.java

```diff
@@ -0,0 +1,370 @@
+     */
+    @Override
+    public ExtendedDnsCacheEntry cache(String hostname, DnsPtrRecord ptrRecord, EventLoop loop) {
+        checkNotNull(hostname, "hostname");
+        checkNotNull(ptrRecord, "ptrRecord");
+        checkNotNull(loop, "loop");
+        DefaultExtendedDnsCacheEntry e = new DefaultExtendedDnsCacheEntry(hostname, ptrRecord);
+        if (maxTtl == 0) {
+            return e;
+        }
+        resolveCache.cache(appendDot(hostname), e, Math.max(minTtl, (int) Math.min(maxTtl, ptrRecord.timeToLive())));
+        return e;
+    }
+
+    /**
+     * Negative caching for failures (or empty reverse lookups).
+     */
+    @Override
+    public DnsCacheEntry cache(String hostname, DnsRecord[] additionals, Throwable cause, EventLoop loop) {
+        checkNotNull(hostname, "hostname");
+        checkNotNull(cause, "cause");
+        checkNotNull(loop, "loop");
+
+        DefaultExtendedDnsCacheEntry e = new DefaultExtendedDnsCacheEntry(hostname, cause);
+        if (negativeTtl == 0 || !emptyAdditionals(additionals)) {
+            return e;
+        }
+
+        resolveCache.cache(appendDot(hostname), e, negativeTtl);
+        return e;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(getClass())
+                .add("minTtl", minTtl)
+                .add("maxTtl", maxTtl)
+                .add("negativeTtl", negativeTtl)
+                .add("cached resolved hostname=", resolveCache.size())
+                .toString();
+    }
+
+    public long getSize() {
+        return resolveCache.size();
+    }
+
+    private enum CacheMetric {
+        SIZE("cacheSize", c -> (Gauge<Long>) c.resolveCache::size),
+        MAX_SIZE("cacheMaxSize", c -> (Gauge<Long>) c::maxSize),
+        EVICTION_COUNT("cacheEvictionCount", c -> (Gauge<Long>) c.resolveCache.stats()::evictionCount),
+        HITS("cacheHits", c -> c.cacheHits),
+        MISSES("cacheMisses", c -> c.cacheMisses);
+
+        String name;
+        Function<CaffeineDnsCache, Metric> metricSupplier;
+
+        CacheMetric(String name, Function<CaffeineDnsCache, Metric> metricSupplier) {
+            this.name = Objects.requireNonNull(name);
+            this.metricSupplier = Objects.requireNonNull(metricSupplier);
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public Metric getMetric(CaffeineDnsCache cache) {
+            return metricSupplier.apply(cache);
+        }
+    }
+
+    public void registerMetrics(MetricRegistry metrics) {
+        for (CacheMetric cm : CacheMetric.values()) {
+            metrics.register(cm.getName(), cm.getMetric(this));
+        }
+    }
+
+    public void unregisterMetrics(MetricRegistry metrics) {
+        for (CacheMetric cm : CacheMetric.values()) {
+            metrics.remove(cm.getName());
+        }
+    }
+
+    private static final class DefaultExtendedDnsCacheEntry implements ExtendedDnsCacheEntry {
```

Comment:
Maybe this should be split in three implementations. On for `hostnameFromPtrRecord`, `address` and `cause` each.

Reply:
I'm not sure if it makes sense to mix forward and reverse lookup. Maybe this should be split into two distinct classes / interfaces and use the same generic cache implementation in the background.

Reply:
I tried splitting this up into different classes, but found that its add a lot of complexity. We then need to introduce something a long the line of a visitor pattern to figure out the type of entries in the container. Can do it, but not sure it's worth it.

Reply:
I think I found a cleaner way to do this while keeping it simple. I'll come back to this.